Tests/MathLib/TestCVode.cpp

```diff
@@ -0,0 +1,337 @@
+{
+	const double pV = y[0];
+	const double C = y[1];
+
+	// std::printf("pV: %14.7g, C: %14.7g\n", pV, C);
+
+	// if (0.0 < pV && pV < 1e-6) pV = 1e-6;
+
+	if (pV >= 1.0)
+	{
+		ydot[1] = ads->get_reaction_rate(pV, T, M, C);
+		ydot[0] = -R * T / M / phi * (1.0 - phi) * rho_SR0 * ydot[1];
+		return true;
+	}
+	else if (pV > 0.0)
+	{
+		ydot[0] = ydot[1] = 0.0;
+		return true;
+	}
+	else
+	{
+		ydot[0] = ydot[1] = 0.0;
+		return false;
+	}
+}
+
+#endif
+
+bool f(const double,
+       MathLib::MappedConstVector<1> const y,
+       MathLib::MappedVector<1> ydot)
+{
+	if (y[0] <= 0.0) return false;
+
+	ydot[0] = -15.0 * y[0];
+	return true;
+}
+
+bool df(const double /*t*/,
+        MathLib::MappedConstVector<1> const y,
+        MathLib::MappedVector<1> /*ydot*/,
+        MathLib::MappedMatrix<1, 1> jac)
+{
+	if (y[0] <= 0.0) return false;
+
+	jac(0, 0) = -15.0;
+
+	// std::printf("jac: %g\n", *jac);
+	return true;
+}
+
+struct ExtraData
+{
+	double value = 15.0;
+};
+
+bool f_extra(const double,
+             MathLib::MappedConstVector<1> const y,
+             MathLib::MappedVector<1> ydot,
+             ExtraData& data)
+{
+	if (y[0] <= 0.0) return false;
+
+	ydot[0] = -data.value * y[0];
+	return true;
+}
+
+#ifdef ZEOLITE
+TEST(MathLibCVodeTest, ZeoliteAdsorption)
+{
+	// initial values
+	const double C = 0.0;
+	const double pV = 100.0;  // Pa
+
+	ads = Adsorption::newInstance(SolidReactiveSystem::Z13XBF_Hauer);
+
+	MathLib::CVodeSolver::ConfigTree config;
+	config.put("linear_multistep_method", "BDF");
+	auto ode_solver = MathLib::createOdeSolver<2>(config);
+
+	ode_solver->setFunction(f_zeolite, nullptr);
+	ode_solver->setTolerance({1.0, 0.01}, 1.0);
+
+	ode_solver->setIC(0.0, {{pV, C}});
+
+	ode_solver->preSolve();
+
+	const double dt = 1e-4;
+
+	for (unsigned i = 1; i <= 10; ++i)
+	{
+		const double time = dt * i;
+
+		ode_solver->solve(time);
+
+		double const* y = ode_solver->getSolution();
+		double time_reached = ode_solver->getTime();
+		auto const y_dot = ode_solver->getYDot(time_reached, y);
+
+		// std::printf("t: %g, pV: %g, C: %g\n", time_reached, y[0], y[1]);
+		std::printf("[ %23.16g, %23.16g, %13.16g ],\n", time_reached, y[0],
+		            y[1]);
+
+		ASSERT_NEAR(time, time_reached, std::numeric_limits<double>::epsilon());
+		// std::printf("time: %g\n", time_reached);
+	}
+}
+#endif
+
+TEST(MathLibCVodeTest, Exponential)
+{
+	// initial values
+	const double y0 = 1.0;
+	const double t0 = 0.0;
+
+	auto tree = boost::property_tree::ptree{};
+	BaseLib::ConfigTree config(tree, "",
+	                           BaseLib::ConfigTree::onerror,
+	                           BaseLib::ConfigTree::onwarning);
+	auto ode_solver = MathLib::createOdeSolver<1>(config);
+
+	ode_solver->setFunction(f, nullptr);
+	ode_solver->setTolerance(1e-8, 1e-6);
+
+	ode_solver->setIC(t0, {{y0}});
+
+	ode_solver->preSolve();
+
+	const double dt = 1e-1;
+
+	for (unsigned i = 1; i <= 10; ++i)
+	{
+		const double time = dt * i;
+
+		ode_solver->solve(time);
```

Comment:
check inside `solve` if time is reached, return error code. Maybe covered by sundials `check_error`